lib/cfg-lexer.c

```diff
@@ -1050,13 +1047,49 @@ cfg_lexer_lex(CfgLexer *self, YYSTYPE *yylval, YYLTYPE *yylloc)
       self->non_pragma_seen = TRUE;
     }
 
-  if (!injected)
+  return TRUE;
+}
+
+int
+cfg_lexer_lex(CfgLexer *self, YYSTYPE *yylval, YYLTYPE *yylloc)
+{
+  /*
+   * NOTE:
+   *
+   * String tokens are allocated by malloc/free and not
+   * g_malloc/g_free, this is significant.  The grammar contains the free()
+   * call, so getting rid of that would require a lot of changes to the
+   * grammar. (on Windows glib, malloc/g_malloc are NOT equivalent)
+   *
+   */
+
+  gint tok;
+  gboolean lex_again;
+  gboolean is_token_injected;
+
+  do
     {
-      if (self->preprocess_suppress_tokens == 0 && self->preprocess_output)
+      is_token_injected = cfg_lexer_consume_next_injected_token(self, &tok, yylval, yylloc);
+
+      if (!is_token_injected)
         {
-          g_string_append_printf(self->preprocess_output, "%s", self->token_text->str);
+          if (cfg_lexer_get_context_type(self) == LL_CONTEXT_BLOCK_CONTENT)
+            cfg_lexer_start_block_state(self, "{}");
+          else if (cfg_lexer_get_context_type(self) == LL_CONTEXT_BLOCK_ARG)
+            cfg_lexer_start_block_state(self, "()");
+
+          tok = cfg_lexer_lex_next_token(self, yylval, yylloc);
+          cfg_lexer_append_preprocessed_output(self, self->token_pretext->str);
         }
+
+      if (!cfg_lexer_process_token(self, tok, yylval, yylloc, &lex_again))
```

Comment:
This function is a little confusing for me. It has a return value, and other bool value via parameter (`relex_required`), and those two variable describes what the next action should be. Those two variable are tangled, but the code does not represents it.
Please if possible couple those variables.

Reply:
OK, I have an idea.

Reply:
Fixed, thanks.